query-engine/schema/src/query_schema.rs

```diff
@@ -177,6 +180,24 @@ impl QuerySchema {
             .capabilities()
             .contains(ConnectorCapability::NativeUpsert)
     }
+
+    /// Sets the database version at runtime to exclude MySQL < 8.0.14 from the "join" relation load strategy.
+    /// Make sure to set this field whenever you intend to execute queries.
+    ///
+    // HACK: We store the database version at runtime to exclude MySQL < 8.0.14 from the "join" relation load strategy.
+    // HACK: This is ugly and out of place. The database version should be statically known through the Prisma Schema.
+    // HACK: We hold this info here because we can't pass it to the connector because they're ARCed.
+    pub fn set_database_version(&mut self, db_version: String) {
+        self.database_version = Some(db_version);
+    }
+
+    // HACK: We retrieve the database version at runtime to exclude MySQL < 8.0.14 from the "join" relation load strategy.
+    // HACK: This is ugly and out of place. The database version should be statically known through the Prisma Schema.
+    // HACK: We hold this info here because we can't pass it to the connector because they're ARCed.
+    // HACK: Please refrain from using this in a different context than RelationLoadStrategy at all cost.
```

Comment:
This makes me think that what we really want to store here is not the database version but whether the database supports joins. This is sort of relevant to the query schema, and it will also allow to avoid parsing and checking the version on every request.

Reply:
It crossed my mind yesterday but also thought:
1. It'd be worse to have no context as to how that boolean is computed (from the perspective of the Query Schema)
2. That field could be misused in the Query Schema builder

Point n°2 is important because if we ever were to use that boolean field in the query schema builder, we'd start desynchronizing the QE's Query Schema and the Prisma Client's Query Schema (since that boolean field is always false when computing DMMF).

In this specific case, it could mean that a user gets a parsing error for the `relationLoadStrategy` not existing as opposed to a proper error explaining that certain MySQL versions don't support the `relationLoadStrategy: join`.

Nonetheless, I think it's cleaner from a code perspective. I have updated it. This comment might give context to future readers. I really hope we, one day, are forced to add a static version field is PSL to get rid of that dirt.

Reply:
> Point n°2 is important because if we ever were to use that boolean field in the query schema builder, we'd start desynchronizing the QE's Query Schema and the Prisma Client's Query Schema (since that boolean field is always false when computing DMMF).

That's a very good point. Maybe it should be `{ Yes, No, UnknownYet }` enum rather than a boolean?

Reply:
Have a look at the latest commit. That's indeed what I've ended up doing, but the enum is derived from the stored boolean in Query Schema. I like the idea of `UnknownYet` which can be changed to `Yes`/`No` depending on the additional info we get from the database version.

It's unclear how we can compute the `UnknownYet` heuristic without getting too specific about connectors though. I'd like to avoid specifically checking for MySQL in the `QuerySchema`. The only abstract heuristic we have is the `ConnectorCapability::CorrelatedSubqueries`, but that won't be enough because SQLite might very well inherit that capability.

Reply:
> Have a look at the latest commit. That's indeed what I've ended up doing, but the enum is derived from the stored boolean in Query Schema. I like the idea of UnknownYet which can be changed to Yes/No depending on the additional info we get from the database version.

Yes, my idea was that the problem that the schema builder could potentially incorrectly use the boolean flag to remove the join strategy is really caused by the fact that we have three distinct states but a boolean only provides two values, so we need a type with three values, like `Option<bool>` or an explicit enum that I suggested. So we would initialize the field with `UnknownYet` (or `None`) at first, and only once we got the version from the database we would replace it with `Supported`/`Unsupported` (or `Some(true)`/`Some(false)`).

Reply:
So previously the value could transition from `false` to `false` or from `false` to `true` (with `false` being ambiguous as it means two different things while building the schema and afterwards), and instead it would always be `UnknownYet` when building the schema and then transition to either `Supported` or `Unsupported`.